### ISSUE TITLE
Transmission can't automatically set upload speed

### ISSUE THREAD

**Issue by ballo** [transmission/transmission#3945]
### What is the issue?

This is a feature that's over 10 years old, so I'm not sure why it isn't more prevalent. I'm always on the road and have to manually set this daily, or use a different client.

### Which application of Transmission?

macOS app

### Which version of Transmission?

3.0.0

**Comment by Coeur** [comment]
You'd like to auto change the speed of Transmission when you're using your Mac in a car? 😕 

Otherwise, in one click, in Transmission 4, after configuring your desired maximum speeds in the Settings for the Turtle mode, you can switch the Turtle mode on and off in the lower left corner of Transmission.

**Comment by ballo** [comment]
I want it to set the upload cap automatically like uTorrent or Vuze

**Comment by Coeur** [comment]
I found a doc of Vuze mentioning "Auto-Speed":
http://wiki.vuze.com/w/UG_Options

Screenshot says:
>Auto-speed automatically adjusts the upload speed limit to avoid overloading the network connection

**Comment by GaryElshaw** [comment]
I'm even more confused. @ballo have you looked at the 'Global bandwidth limits' in 'Bandwidth'? You can establish a bandwidth criteria for *all* right there.

**Comment by Coeur** [comment]
From the screenshot of Vuze Auto-Speed (beta), instead of setting a hard limit, it's configured as a percentage of your available network bandwidth (like "80%" of network capacities)

**Comment by GaryElshaw** [comment]
¯\_(ツ)_/¯ I get it, but...

**Comment by ballo** [comment]
It shares fine, but without sacrificing download speed. If anything it encourages sharing because it doesn't get in the way of downloading

**Comment by ckerr** [comment]
I think I'm misreading this because Coeur and Gary seem to understand the request but I don't.

> What is the issue?
>
> This is a feature that's over 10 years old, so I'm not sure why it isn't more prevalent. I'm always on the road and have to manually set this daily, or use a different client.

That's not really a description.

> I want it to set the upload cap automatically like uTorrent or Vuze

You can already cap your upload speed based on time of day, so I guess that's not what you're requesting. Could you rephrase what you're looking for?

**Comment by Coeur** [comment]
@ckerr ballo wants the Vuze Auto-Speed feature which caps the download to a percentage of your max available bandwidth. Say that you set the cap to 80%, then you'll download at max 800KB/s if your available download bandwidth on your computer is presently 1000 KB/s and you'll upload at max 80KB/s if your available upload bandwidth on your computer is presently 100 KB/s.

**Comment by ballo** [comment]
This is how uTorrent does it:
https://forum.utorrent.com/topic/279-automatic-bandwidth-restriction/

VUZE looks similar:
https://wiki.vuze.com/w/Auto_Speed_Beta#Algorithm_details

**Comment by abubaca4** [comment]
Look like this feature implemented in [this](https://github.com/BiglySoftware/BiglyBT/blob/d80c56ee44d44a4ff5a6b5df5a9666a0700986d8/core/src/com/biglybt/core/speedmanager/impl/v1/SpeedManagerAlgorithmProviderV1.java) [files](https://github.com/BiglySoftware/BiglyBT/blob/d80c56ee44d44a4ff5a6b5df5a9666a0700986d8/core/src/com/biglybt/core/speedmanager/impl/SpeedManagerImpl.java) of BiglyBT

**Comment by abubaca4** [comment]
I don't know transmission code but I whant to give idea how it shoud work in my mind.
Need one more speed limit that adjust to transfer delay.
So we have target delay from settings.json(in utorrent default is 100ms in net.utp_target_delay, in vuze default is 200ms).
We can get two timestamps when we send utp package and receive confirmation. From this we can take ms taken to transfer.
Good idea is to take median(or something like this) not average because it can be peer from other continent.
Store all delay in vector is not good idea, it's slow. So we can make array like `std::array<uint, taget_delay*2>` full of zero. And for all delay we can do +1 to cell with index count of ms we get. If delay bigger than `taget_delay*2` we use last cell. From this array we can calculate median or just take index of the biggest one cell.
In some interval we get median of array and clear array. Than if delay lower than limit we increase speed limit or do nothing(if speed limit reached global speed limit or not reached current limit). If delay bigger than limit we decrease speed limit.
Interval and steps of speed increase/decrease can be set in settings.json. Default value we can get from vuze where it work a long time.

While I seed in windows I use utorrent 2.2.1 with feature like this. And it very helpful in connection without slow upload like adsl docsis mobile etc. Without it ping can increase to 300-500ms and we need manually set limit(that can change with load from other abbonents) or stop seeding to normal use internet.
